File: bt/binary_tree.c

#include <stdio.h>
#include <stdlib.h>
#include "binary_tree.h"


typedef struct BtNode
{
    int value;
    struct BtNode *left;
    struct BtNode *right;
} BtNode;

struct BinaryTree
{
    BtNode *root;
};


BtNode *node_construct(int value, BtNode *left, BtNode *right)
{
    BtNode *node = (BtNode*) calloc(1, sizeof(BtNode));

    node->value = value;
    node->left = left;
    node->right = right;

    return node;
}


void node_destroy(BtNode *node)
{
    free(node);
}


BinaryTree *binary_tree_construct()
{
    BinaryTree *bt = (BinaryTree *) calloc(1, sizeof(BinaryTree));
    bt->root = NULL;
    return bt;
}


void binary_tree_add(BinaryTree *bt, int value)
{
    BtNode *new_node = node_construct(value, NULL, NULL);

    if (bt->root == NULL)
    {
        bt->root = new_node;
    }
    else
    {
        BtNode *node = bt->root;

        while (node != NULL)
        {
            if (value == node->value)
            {
                node_destroy(new_node);
                return;
            }
            else if (value < node->value)
            {
                if (node->left == NULL)
                {
                    node->left = new_node;
                    return;
                }
                else
                    node = node->left;
            }
            else
            {
                if (node->right == NULL)
                {
                    node->right = new_node;
                    return;
                }
                else
                    node = node->right;
            }
        }
    }
}


int binary_tree_exists(BinaryTree *bt, int val)
{
    BtNode *n = bt->root;

    while (n != NULL)
    {
        if (val == n->value) return 1;
        else if (val < n->value) n = n->left;
        else n = n->right;
    }

    return 0;
}


int binary_tree_min(BinaryTree *bt)
{
    if (bt->root == NULL)
        exit(printf("Error: empty tree\n"));

    BtNode *n = bt->root;

    while (n->left != NULL)
        n = n->left;

    return n->value;
}


int binary_tree_max(BinaryTree *bt)
{
    if (bt->root == NULL)
        exit(printf("Error: empty tree\n"));

    BtNode *node = bt->root;

    while (node->right != NULL)
        node = node->right;

    return node->value;
}


void _destroy_node_recursive(BtNode *node)
{
    if (node == NULL)
        return;

    _destroy_node_recursive(node->left);
    _destroy_node_recursive(node->right);
    node_destroy(node);
}


void binary_tree_destroy(BinaryTree *bt)
{
    _destroy_node_recursive(bt->root);
    free(bt);
}

File: bt/BinaryTree.java
package bt;
import java.util.NoSuchElementException;

public class BinaryTree {
    private BtNode root;

    public BinaryTree() {
        this.root = null;
    }

    public void binary_tree_add(int value) {
        BtNode novo_n = new BtNode(value, null, null);

        if (this.root == null) {
            this.root = novo_n;
        } else {
            BtNode node = this.root;

            while (node != null) {
                if (value == node.get_value()) {
                    break;
                } else if (value < node.get_value()) {
                    if (node.get_left() == null) {
                        node.set_left(novo_n);
                        break;
                    } else
                        node = node.get_left();
                } else {
                    if (node.get_right() == null) {
                        node.set_right(novo_n);
                        return;
                    } else
                        node = node.get_right();

                }
            }
        }
    }

    public int binary_tree_exists(int val) {
        BtNode n = this.root;

        while (n != null) {
            if (val == n.get_value())
                return 1;
            else if (val < n.get_value())
                n = n.get_left();
            else
                n = n.get_right();
        }

        return 0;
    }

    public int binary_tree_min() {
        if (this.root == null)
            throw new NoSuchElementException("Error: empty tree\n");

        BtNode n = this.root;

        while (n.get_left() != null)
            n = n.get_left();

        return n.get_value();
    }

    public int binary_tree_max() {
        if (this.root == null)
            throw new NoSuchElementException("Error: empty tree\n");

        BtNode n = this.root;

        while (n.get_right() != null)
            n = n.get_right();

        return n.get_value();
    }

}